### PR TITLE
Account history

### PR DESCRIPTION
**Pull request by tovebr**

### CODE REVIEW

File: Account/History.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	timestamp: isoly.DateTime
+	property: "status" | "rails"
+	to: string
+}
+
+export namespace History {
+	export type Property = typeof Property.values[number]
+	export namespace Property {
+		export const values = ["status", "rails", "counterparts", "rules"] as const
+	}
```
Comment (SiFalorSum) [inline Account/History.ts:14]: Add Property isly "type" in namespace

File: Account/History.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	export const type = isly.object<History>({
+		timestamp: isly.string<Property>(Property.values),
```
Comment (SiFalorSum) [inline Account/History.ts:16]: this is not a timestamp type

File: Account/History.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { isoly } from "isoly"
+import { isly } from "isly"
+
+export type History = {
+	timestamp: isoly.DateTime
+	property: "status" | "rails"
```
Comment (SiFalorSum) [inline Account/History.ts:6]: Use Property type here

File: Account/Status.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+import { isly } from "isly"
+
+export type Status = {
+	mode: "active" | "frozen" | "closed"
+	reason?: "overdraft" | "other" | string
```
Comment (SiFalorSum) [inline Account/Status.ts:5]: Suggestion: `reason?: "overdraft" | "other" | (string & Record<never, never>)` to fool typescript and get type suggestions, now Typescript will simplify this type to just `string`

File: Account/Status.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import { isly } from "isly"
+
+export type Status = {
+	mode: "active" | "frozen" | "closed"
```
Comment (SiFalorSum) [inline Account/Status.ts:4]: Use `Mode` type here

File: Account/Status.ts (hunk 4 above)
Comment (SiFalorSum) [reply Account/Status.ts:5]: Also, put this type in the namespace

File: Account/index.ts [hunk 6]
```diff
@@ -14,8 +16,20 @@ export interface Account extends Account.Creatable {
 	counterparts?: Record<string, Rail.Address>
 	key?: string
 	rules?: Rule[]
+	status: AccountStatus
+	history: AccountHistory[]
```
Comment (SiFalorSum) [inline Account/index.ts:20]: remove `history` from interface

**Review by SiFalorSum — CHANGES_REQUESTED** [review]
See comments

File: Account/History.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+	to: string
```
[the remark this answers is not in the document]
Comment (marcushanses) [reply Account/History.ts:7]: We could also do, Status | Rails[] | string

**Review by marcushanses — CHANGES_REQUESTED** [review]

File: Account/index.ts [hunk 8]
```diff
@@ -14,8 +16,13 @@ export interface Account extends Account.Creatable {
 	counterparts?: Record<string, Rail.Address>
 	key?: string
 	rules?: Rule[]
+	status: Status
 }
 export namespace Account {
+	export type Legacy = Omit<Account, "status">
+	export function fromLegacy(maybeLegacy: Legacy | Account, status?: Status): Account {
+		return { ...maybeLegacy, status: status ?? ("status" in maybeLegacy ? maybeLegacy.status : { mode: "active" }) }
+	}
```
Comment (marcushanses) [inline Account/index.ts:25]: Remember to add history to accounts and if it's not optional we should add it to the fromLegacy function.

**Review by marcushanses — CHANGES_REQUESTED** [review]

File: Account/History.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+export namespace History {
+	export type Property = typeof Property.values[number]
+	export namespace Property {
+		export const values = ["status", "rails", "counterparts", "rules"] as const
```
[the remark this answers is not in the document]
Comment (tovebr) [reply Account/History.ts]: Added "created" to Property!

File: Account/index.ts (hunk 8 above)
Comment (tovebr) [reply Account/index.ts:25]: History is not supposed to be on account!

File: Account/History.ts (hunk 1 above)
Comment (tovebr) [reply Account/History.ts:15]: +

File: Account/History.ts (hunk 2 above)
Comment (tovebr) [reply Account/History.ts]: +

File: Account/History.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import { isoly } from "isoly"
+import { isly } from "isly"
+
+export type History = {
+	timestamp: isoly.DateTime
+	property: "status" | "rails" | "counterparts" | "rules" | "created" | "updated"
```
Comment (SiFalorSum) [inline Account/History.ts:6]: Use `Property` type

File: Account/index.ts [hunk 11]
```diff
@@ -25,11 +32,13 @@ export namespace Account {
...
 		rules: Rule.type.array().optional(),
+		status: Status.type,
 	})
 	export const is = type.is
 	export const flaw = type.flaw
 	export function isIdentifier(value: cryptly.Identifier | any): value is cryptly.Identifier {
 		return cryptly.Identifier.is(value, 8)
 	}
 	export import Creatable = AccountCreatable
+	export import History = AccountHistory
```
Comment (SiFalorSum) [inline Account/index.ts:43]: Would be nice to have `Status` exported as well

**Pull request closed without merging**